File: kernel-pipe-throughput/bench.zig
const std = @import("std");
const posix = std.posix;

const TOTAL_BYTES: u64 = 10 * 1024 * 1024 * 1024; // 10 GB
const BUFFER_SIZE: usize = 64 * 1024; // 64 KB

pub fn main() !void {
    const pipe_result = try posix.pipe();
    const read_fd = pipe_result[0];
    const write_fd = pipe_result[1];

    const pid = try posix.fork();

    if (pid == 0) {
        // Child: Reader
        posix.close(write_fd);

        var buffer: [BUFFER_SIZE]u8 = undefined;
        var total_received: u64 = 0;
        var checksum: u8 = 0;

        while (total_received < TOTAL_BYTES) {
            const n = posix.read(read_fd, &buffer) catch |err| {
                if (err == error.WouldBlock) continue;
                break;
            };
            if (n == 0) break;

            for (buffer[0..n]) |byte| {
                checksum ^= byte;
            }
            total_received += n;
        }

        const stdout = std.io.getStdOut().writer();
        try stdout.print("CHECK:{x:0>2}\n", .{checksum});

        posix.close(read_fd);
        posix.exit(0);
    } else {
        // Parent: Writer
        posix.close(read_fd);

        var buffer: [BUFFER_SIZE]u8 = undefined;
        // Fill buffer with sequential pattern
        for (0..BUFFER_SIZE) |i| {
            buffer[i] = @truncate(i);
        }

        var timer = try std.time.Timer.start();
        var total_sent: u64 = 0;

        while (total_sent < TOTAL_BYTES) {
            const n = posix.write(write_fd, &buffer) catch |err| {
                if (err == error.WouldBlock) continue;
                break;
            };
            total_sent += n;
        }

        const elapsed_ns = timer.read();
        const elapsed_ms: f64 = @as(f64, @floatFromInt(elapsed_ns)) / 1_000_000.0;
        const throughput_gb_sec = (@as(f64, @floatFromInt(TOTAL_BYTES)) / 1024.0 / 1024.0 / 1024.0) / (elapsed_ms / 1000.0);

        posix.close(write_fd);
        _ = posix.waitpid(pid, 0);

        const stdout = std.io.getStdOut().writer();
        try stdout.print("elapsed_ms={d:.3} throughput_gb_sec={d:.3}\n", .{ elapsed_ms, throughput_gb_sec });
    }
}

File: kernel-pipe-throughput/bench.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <time.h>
#include <sys/wait.h>
#include <stdint.h>

#define TOTAL_BYTES (10ULL * 1024 * 1024 * 1024) // 10 GB
#define BUFFER_SIZE (64 * 1024)                  // 64 KB

static inline double now_ms(void) {
    struct timespec ts;
    clock_gettime(CLOCK_MONOTONIC, &ts);
    return (double)ts.tv_sec * 1000.0 + (double)ts.tv_nsec / 1.0e6;
}

int main() {
    int pipe_fds[2];
    if (pipe(pipe_fds) == -1) {
        perror("pipe");
        return 1;
    }

    pid_t pid = fork();
    if (pid == -1) {
        perror("fork");
        return 1;
    }

    if (pid == 0) { // Child: Reader
        close(pipe_fds[1]);
        uint8_t *buffer = malloc(BUFFER_SIZE);
        uint64_t total_received = 0;
        uint8_t checksum = 0;

        while (total_received < TOTAL_BYTES) {
            ssize_t n = read(pipe_fds[0], buffer, BUFFER_SIZE);
            if (n <= 0) break;
            for (ssize_t i = 0; i < n; i++) checksum ^= buffer[i];
            total_received += n;
        }
        
        printf("CHECK:%02x\n", checksum);
        free(buffer);
        close(pipe_fds[0]);
        exit(0);
    } else { // Parent: Writer
        close(pipe_fds[0]);
        uint8_t *buffer = malloc(BUFFER_SIZE);
        // Fill buffer with some data
        for (int i = 0; i < BUFFER_SIZE; i++) buffer[i] = (uint8_t)i;

        double start = now_ms();
        uint64_t total_sent = 0;

        while (total_sent < TOTAL_BYTES) {
            ssize_t n = write(pipe_fds[1], buffer, BUFFER_SIZE);
            if (n <= 0) break;
            total_sent += n;
        }

        double elapsed_ms = now_ms() - start;
        printf("elapsed_ms=%.3f throughput_gb_sec=%.3f\n", 
               elapsed_ms, (TOTAL_BYTES / 1024.0 / 1024.0 / 1024.0) / (elapsed_ms / 1000.0));

        close(pipe_fds[1]);
        wait(NULL);
        free(buffer);
    }

    return 0;
}
